pkg/rbd/rbd_util.go

```diff
@@ -467,141 +1034,252 @@ func protectSnapshot(pOpts *rbdSnapshot, adminID string, credentials map[string]
+			": (2) No such file or directory") {
+			return imgInfo, ErrImageNotFound{imageName, err}
+		}
+		return imgInfo, err
 	}
 
-	image := pVolOpts.VolName
-	snapID := pSnapOpts.SnapID
-
-	key, err := getRBDKey(pVolOpts.ClusterID, adminID, credentials)
+	err = json.Unmarshal(stdout, &imgInfo)
 	if err != nil {
-		return err
+		klog.Errorf("failed to parse JSON output of image info (%s): (%s)",
+			poolName+"/"+imageName, err)
+		return imgInfo, fmt.Errorf("unmarshal failed: %+v.  raw buffer response: %s",
+			err, string(stdout))
 	}
-	klog.V(4).Infof("rbd: clone %s using mon %s, pool %s", image, mon, pVolOpts.Pool)
-	args := []string{"clone", pSnapOpts.Pool + "/" + pSnapOpts.VolName + "@" + snapID, pVolOpts.Pool + "/" + image, "--id", adminID, "-m", mon, "--key=" + key}
 
-	output, err = execCommand("rbd", args)
+	return imgInfo, nil
+}
+
+// snapInfo strongly typed JSON spec for snap ls rbd output
+type snapInfo struct {
+	ID        int64  `json:"id"`
+	Name      string `json:"name"`
+	Size      int64  `json:"size"`
+	Timestamp string `json:"timestamp"`
+}
+
+// ErrSnapNotFound is returned when snap name passed is not found in the list of snapshots for the
+// given image
+type ErrSnapNotFound struct {
+	snapName string
+	err      error
+}
+
+func (e ErrSnapNotFound) Error() string {
+	return e.err.Error()
+}
+
+/*
+getSnapInfo queries rbd about the snapshots of the given image and returns its metadata, and
+returns ErrImageNotFound if provided image is not found, and ErrSnapNotFound if provided snap
+is not found in the images snapshot list
+*/
+func getSnapInfo(monitors, adminID, key, poolName, imageName, snapName string) (snapInfo, error) {
+	// rbd --format=json snap ls [image-spec]
 
+	var (
+		snpInfo snapInfo
+		snaps   []snapInfo
+	)
+
+	stdout, _, err := util.ExecCommand(
+		"rbd",
+		"-m", monitors,
+		"--id", adminID,
+		"--key="+key,
+		"-c", util.CephConfigPath,
+		"--format="+"json",
+		"snap", "ls", poolName+"/"+imageName)
 	if err != nil {
-		return errors.Wrapf(err, "failed to restore snapshot, command output: %s", string(output))
+		klog.Errorf("failed getting snap (%s) information from image (%s): (%s)",
+			snapName, poolName+"/"+imageName, err)
+		if strings.Contains(string(stdout), "rbd: error opening image "+imageName+
+			": (2) No such file or directory") {
+			return snpInfo, ErrImageNotFound{imageName, err}
+		}
+		return snpInfo, err
 	}
 
-	return nil
+	err = json.Unmarshal(stdout, &snaps)
+	if err != nil {
+		klog.Errorf("failed to parse JSON output of image snap list (%s): (%s)",
+			poolName+"/"+imageName, err)
+		return snpInfo, fmt.Errorf("unmarshal failed: %+v. raw buffer response: %s",
+			err, string(stdout))
+	}
+
+	for _, snap := range snaps {
+		if snap.Name == snapName {
+			return snap, nil
+		}
+	}
+
+	return snpInfo, ErrSnapNotFound{snapName, fmt.Errorf("snap (%s) for image (%s) not found",
+		snapName, poolName+"/"+imageName)}
 }
```

Comment:
This file looks huge to me, can we split it up if possible?

Reply:
Split out the functions that deal with keys and objects, IOW maintain the journal. Should now be reasonable in size and logically split.

FYI, further as the CephFS implementation is done, the voljournal should move to util instead of being retained in the rbd directory.